### PR TITLE
fix: Produce valid CRDs containing flattened untagged enums

### LINKED ISSUE #1941: Flattened untagged enums lead to invalid CRDs
### Current and expected behavior

> [!TIP]
> For anyone implementing the fix, the code used below was written by @sbernauer (and slightly modified by me). Please add attributions to the commit(s) if you copy/paste it:
> ```
> Co-authored-by: Sebastian Bernauer <sebastian.bernauer@stackable.tech>
> Co-authored-by: Nick Larsen <NickLarsenNZ@users.noreply.github.com>
> ```

## Current bahaviour:

```rust
#[derive(CustomResource, Serialize, Deserialize, Debug, Clone, JsonSchema)]
#[kube(group = "clux.dev", version = "v1", kind = "FlattenedUntaggedEnumTest")]
struct FlattenedUntaggedEnumTestSpec {
    foo: FlattenedUntaggedEnum,
}

/// An untagged enum with a nested enum inside
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
#[serde(untagged)]
enum UntaggedEnum {
    /// Used in case the `one` field of type [`u32`] is present
    /// This should not appear in the schema because the "variant" disappears
    /// and this comment cannot pertain to all fields within the struct variant.
    A { one: String },
    /// Used in case the `two` field of type [`NormalEnum`] is present
    /// This should not appear in the schema because the "variant" disappears
    /// and this comment cannot pertain to all fields within the struct variant.
    B { two: NormalEnum, three: String },
    /// Used in case no fields are present
    C {},
}

/// Put a [`UntaggedEnum`] behind `#[serde(flatten)]`
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
struct FlattenedUntaggedEnum {
    #[serde(flatten)]
    inner: UntaggedEnum,
}
```

<details><summary>Invalid CRD</summary>

```json
{
  "apiVersion": "apiextensions.k8s.io/v1",
  "kind": "CustomResourceDefinition",
  "metadata": {
    "name": "flatteneduntaggedenumtests.clux.dev"
  },
  "spec": {
    "group": "clux.dev",
    "names": {
      "categories": [],
      "kind": "FlattenedUntaggedEnumTest",
      "plural": "flatteneduntaggedenumtests",
      "shortNames": [],
      "singular": "flatteneduntaggedenumtest"
    },
    "scope": "Cluster",
    "versions": [
      {
        "additionalPrinterColumns": [],
        "name": "v1",
        "schema": {
          "openAPIV3Schema": {
            "description": "Auto-generated derived type for FlattenedUntaggedEnumTestSpec via `CustomResource`",
            "properties": {
              "spec": {
                "properties": {
                  "foo": {
                    "anyOf": [
                      {
                        "required": [
                          "one"
                        ]
                      },
                      {
                        "required": [
                          "three",
                          "two"
                        ]
                      },
                      {
                        "description": "Used in case no fields are present",
                        "type": "object"
                      }
                    ],
                    "description": "Put a [`UntaggedEnum`] behind `#[serde(flatten)]`",
                    "properties": {
                      "one": {
                        "description": "Used in case the `one` field of type [`u32`] is present\nThis should not appear in the schema because the \"variant\" disappears\nand this comment cannot pertain to all fields within the struct variant.",
                        "type": "string"
                      },
                      "three": {
                        "type": "string"
                      },
                      "two": {
                        "description": "A very simple enum with unit variants",
                        "enum": [
                          "C",
                          "D",
                          "A",
                          "B"
                        ],
                        "type": "string"
                      }
                    },
                    "type": "object"
                  }
                },
                "required": [
                  "foo"
                ],
                "type": "object"
              }
            },
            "required": [
              "spec"
            ],
            "title": "FlattenedUntaggedEnumTest",
            "type": "object"
          }
        },
        "served": true,
        "storage": true,
        "subresources": {}
      }
    ]
  }
}
```

</details>

When applied to Kubernetes, the following errors are given:

```
The CustomResourceDefinition "flatteneduntaggedenumtests.clux.dev" is invalid: 
* spec.validation.openAPIV3Schema.properties[spec].properties[foo].anyOf[2].description: Forbidden: must be empty to be structural
* spec.validation.openAPIV3Schema.properties[spec].properties[foo].anyOf[2].type: Forbidden: must be empty to be structural
```

## Expected behaviour

<details><summary>This test should pass</summary>

### Notice the empty `{}` under the `anyOf`.

```rust
#[test]
fn flattened_untagged_enum() {
    assert_json_eq!(
        FlattenedUntaggedEnumTest::crd(),
        json!(
          {
            "apiVersion": "apiextensions.k8s.io/v1",
            "kind": "CustomResourceDefinition",
            "metadata": {
              "name": "flatteneduntaggedenumtests.clux.dev"
            },
            "spec": {
              "group": "clux.dev",
              "names": {
                "categories": [],
                "kind": "FlattenedUntaggedEnumTest",
                "plural": "flatteneduntaggedenumtests",
                "shortNames": [],
                "singular": "flatteneduntaggedenumtest"
              },
              "scope": "Cluster",
              "versions": [
                {
                  "additionalPrinterColumns": [],
                  "name": "v1",
                  "schema": {
                    "openAPIV3Schema": {
                      "description": "Auto-generated derived type for FlattenedUntaggedEnumTestSpec via `CustomResource`",
                      "properties": {
                        "spec": {
                          "properties": {
                            "foo": {
                              "anyOf": [
                                {
                                  "required": [
                                    "one"
                                  ]
                                },
                                {
                                  "required": [
                                    "three",
                                    "two"
                                  ]
                                },
                                {}
                              ],
                              "description": "Put a [`UntaggedEnum`] behind `#[serde(flatten)]`",
                              "properties": {
                                "one": {
                                  "type": "string"
                                },
                                "two": {
                                  "description": "A very simple enum with unit variants",
                                  "enum": [
                                    "C",
                                    "D",
                                    "A",
                                    "B"
                                  ],
                                  "type": "string"
                                },
                                "three": {
                                  "type": "string"
                                }
                              },
                              "type": "object"
                            }
                          },
                          "required": [
                            "foo"
                          ],
                          "type": "object"
                        }
                      },
                      "required": [
                        "spec"
                      ],
                      "title": "FlattenedUntaggedEnumTest",
                      "type": "object"
                    }
                  },
                  "served": true,
                  "storage": true,
                  "subresources": {}
                }
              ]
            }
          }
        )
    );
}
```

Which fails with:

```
json atom at path ".spec.versions[0].schema.openAPIV3Schema.properties.spec.properties.foo.properties.one.description" is missing from rhs

json atom at path ".spec.versions[0].schema.openAPIV3Schema.properties.spec.properties.foo.anyOf[2].type" is missing from rhs

json atom at path ".spec.versions[0].schema.openAPIV3Schema.properties.spec.properties.foo.anyOf[2].description" is missing from rhs
```

</details>

### Possible solution

This is another one that was fixed properly in https://github.com/kube-rs/kube/pull/1839.
Specifically:

https://github.com/stackabletech/kube-rs/blob/71707abf1109bad01bbf2c46e3b161846049767a/kube-core/src/schema/transform_properties.rs#L89-L95

### Additional context

_No response_

### Environment

N/A

### Configuration and features

_No response_

### YAML

_No response_

### Affected crates

_No response_

### Would you like to work on fixing this bug?

yes

### PR DESCRIPTION
**Pull request by NickLarsenNZ**
Fixes #1941

> [!NOTE]
> This is a minimal fix to make review easier and close the issue.
> A follow up PR is intended to replace the properties transformation into a proper transformer like in #1839.

- 1a40d62: Add a test for flattened untagged enums.
- 2caf34b: Ensure `description` and `type` fields are dropped on untagged variants without fields.
- 596bb93: Stop untagged variant descriptions being pushed into properties (which are for the fields)

### CODE REVIEW

File: kube-derive/tests/crd_complex_enum_test.rs [hunk 1]
```diff
@@ -43,6 +43,34 @@ enum ComplexEnum {
...
+    /// Used in case the `two` field of type [`NormalEnum`] is present
+    ///
+    /// This should not appear in the schema because the "variant" disappears
+    /// and this comment cannot pertain to all fields within the struct variant.
+    B { two: NormalEnum, three: String },
+    /// Used in case no fields are present
+    ///
+    /// This should not appear in the schema because the "variant" disappears
+    /// and this comment cannot pertain to all fields within the struct variant.
+    C {},
```
Comment (doxxx93) [inline kube-derive/tests/crd_complex_enum_test.rs:64]: Consider adding a newtype variant like `D(String)` — it produces `{"type": "string"}` (vs `{"type": "object"}` from `C {}`), which also hits the `else if` branch but with a different schema shape.

Comment (NickLarsenNZ) [reply kube-derive/tests/crd_complex_enum_test.rs:64]: Good point! I will push a fix shortly for tuple variants.

Comment (NickLarsenNZ) [reply kube-derive/tests/crd_complex_enum_test.rs:64]: Actually, on second thought, I think that is invalid for an untagged enum.

An untagged enum can either be distinct sets of fields (eg: A and B), or the empty set (eg: C).

I can't imagine what `D(OtherType)` or `D(OtherType1, OtherType2)` should be.

For the first, if it is a struct, you could argue that the struct fields are used. The second seems invalid (or at best, very fragile and prone to accidents).

I'll dig a bit further.

Comment (NickLarsenNZ) [reply kube-derive/tests/crd_complex_enum_test.rs:64]: Ok I tested a bunch of variations:

```rust
    // {
    //   "type": "string" // currently erased
    // }
    D(String),

    // {
    //   "enum": [
    //     null
    //   ],
    //   "nullable": true
    // }
    E(()),

    // {
    //   "items": [
    //     {
    //       "type": "string"
    //     },
    //     {
    //       "enum": [
    //         null
    //       ],
    //       "nullable": true
    //     }
    //   ],
    //   "maxItems": 2,
    //   "minItems": 2,
    //   "type": "array"
    // }
    F(String, ()),

    // {
    //   "description": "A very simple enum with unit variants",
    //   "enum": [
    //     "C",
    //     "D",
    //     "A",
    //     "B"
    //   ],
    //   "type": "string"
    // }
    G(NormalEnum),

    // this is fine (struct fields appear correctly)
    // {
    //   "required": [
    //     "foo"
    //   ]
    // }
    H(NormalEnumTestSpec),
```

I think at least `type: "string"` is invalid for untagged tuple variants of any length (`D` and `F` above). But I also think `D` - `G` are invalid, and `H` is fine.

---

I propose we panic on these conditions for untagged enums:
- `type` != object
- `items` is not None (tuple variant of arity greater than 1)
- `enum` is not None

Comment (NickLarsenNZ) [reply kube-derive/tests/crd_complex_enum_test.rs:64]: I'm also happy for this to be merged as-is, and a separate issue can be raised to discuss my previous comment.

Let me know what you prefer.

Comment (doxxx93) [reply kube-derive/tests/crd_complex_enum_test.rs:64]: Thanks for the thorough investigation! I agree that the tuple variant validation is a separate concern from this PR.

**Comment by doxxx93** [comment]
Sorry for the ping @clux — I've verified the normal behavior through the tests. One thing I noticed: the new else if branch (removing type/metadata from variants without fields) runs for both one_of (tagged) and any_of (untagged) paths, even though the comment says "untagged variants." Is this fine?

Could you take a look when you get a chance?

**Comment by NickLarsenNZ** [comment]
> One thing I noticed: the new else if branch (removing type/metadata from variants without fields) runs for both one_of (tagged) and any_of (untagged) paths, even though the comment says "untagged variants." Is this fine?

No this isn't fine. I will fix that (by only doing it when `push_description_to_property == true` - maybe rename the variable).
This PR is only dealing with untagged enums and should not change anything else.

I didn't see any test failing. Can you share how you tested that?
I tested on one of our CRDs and didn't see a regression, so I'd like to ensure that path is covered.

**Comment by doxxx93** [comment]
@NickLarsenNZ I didn't find a specific failing test case — it was more of a code review observation that the else if branch could theoretically run in the one_of (tagged) path as well. The existing tests with tagged enums pass fine, but I wasn't confident there isn't an uncovered edge case.

File: kube-derive/tests/crd_complex_enum_test.rs (hunk 1 above)
Comment (NickLarsenNZ) [reply kube-derive/tests/crd_complex_enum_test.rs:64]: If I understood correctly, I will leave this PR as-is?

**Comment by NickLarsenNZ** [comment]
@doxxx93 I see what you mean.

These extra fields need to go regardless of oneOf or anyOf, as they are not accepted by Kubernetes.

My original implementation does this (and only preserves the descriptions in the case of a oneOf by moving them into the applicable property).
For the untagged variants, the doc-comment on the variant shouldn't be pushed into the properties because the comment is for the variant and not the field(s) within.

See here: https://github.com/stackabletech/kube-rs/blob/71707abf1109bad01bbf2c46e3b161846049767a/kube-core/src/schema/transform_properties.rs#L89-L95

I would keep it as-is for now, and raise a separate refactor PR based on the properties hoisting in #1839 which is easier to test/debug IMO - if you're ok with that.

File: kube-core/src/schema.rs [hunk 2]
```diff
@@ -499,6 +500,7 @@ fn hoist_subschema_properties(
     subschemas: &mut Vec<Schema>,
     common_obj: &mut Option<Box<ObjectValidation>>,
     instance_type: &mut Option<SingleOrVec<InstanceType>>,
+    push_description_to_property: bool,
```
Comment (clux) [inline kube-core/src/schema.rs:503]: this signature is getting a little awkard and is probably another refactoring candidate for refactoring later.

File: kube-core/src/schema.rs [hunk 3]
```diff
@@ -545,6 +549,19 @@ fn hoist_subschema_properties(
...
+            metadata: variant_metadata,
+            instance_type: variant_type,
+            enum_values: None,
+            subschemas: None,
+            array: None,
+            object: None,
+            ..
+        }) = variant
+        {
+            // This removes the type/description from untagged variants without fields.
```
Comment (clux) [inline kube-core/src/schema.rs]: only nit here is that this comment is put on the bit that is obvious, not why the above `if` match refers to an untagged variant (and from doxx's comments it sounds like it can refer to more than one thing).

a little clarifying on the if condition here would be preferable.

File: kube-core/src/schema.rs (hunk 2 above)
Comment (NickLarsenNZ) [reply kube-core/src/schema.rs:503]: Yes, this was properly handled in the original refactor: https://github.com/stackabletech/kube-rs/blob/71707abf1109bad01bbf2c46e3b161846049767a/kube-core/src/schema/transform_properties.rs#L50-L57

File: kube-core/src/schema.rs (hunk 3 above)
Comment (NickLarsenNZ) [reply kube-core/src/schema.rs]: I'll move the description and clarify what I can (hopefully today).
Again, this does need a proper refactor, which you said you didn't want done as part of fixes. So I'm deferring as much of that work as possible to produce the minimal fix.

Comment (NickLarsenNZ) [reply kube-core/src/schema.rs]: I moved the comment up, and mentioned that it affects both oneOf and anyOf subschemas: 1d0eb38d9d3afafc252e0b915124e32434be0142

**Comment by clux** [comment]
Thank you!
